### PR TITLE
MemoryPacking: Handle non-empty trapping segments

### PR DESCRIPTION
**Pull request by kripken**
Followup to #6243 which handled empty ones.

### CODE REVIEW

**Review by tlively — COMMENTED** [review]
Would it be worth continuing to split these possibly-trapping active segments, but ensure that their last chunks are not elided? Then the initialization of the last chunk (or a previous chunk) would trap iff the original segment would have trapped. Or would it be a problem that this would cause partial initialization to become visible before the trap? I always forget whether we care about side effects before a trap.

**Comment by kripken** [comment]
Interesting idea! I hadn't thought of that.

I think this would be technically incorrect because of partial initialization. At least that was the case in the bulk memory proposal ("bounds checks are done before any writes") - I'm not sure offhand where to read the spec for active segments, but surely we didn't introduce a difference there..?

However it might be an incorrectness we can live with - we already live with reordering traps, for example (which have different stack traces, which is observable). But I think it would be more complicated to do this idea - we'd need to track split-off parts of the original segment etc., and I think for basically no benefit - no real-world code has trapping segments (I hope).

File: test/lit/passes/memory-packing_traps.wast [hunk 1]
```diff
@@ -0,0 +1,96 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+
+;; RUN: foreach %s %t wasm-opt --memory-packing -all --zero-filled-memory      -S -o - | filecheck %s
+;; RUN: foreach %s %t wasm-opt --memory-packing -all --zero-filled-memory -tnh -S -o - | filecheck %s --check-prefix=TNH__
+
+(module
+  ;; We should not optimize out a segment that will trap, as that is an effect
+  ;; we need to preserve (unless TrapsNeverHappen).
+  (memory 1 2 shared)
+  (data (i32.const -1) "\00")
```
Comment (tlively) [inline test/lit/passes/memory-packing_traps.wast:10]: You can give these names to make the output more readable.

**Review by tlively — APPROVED** [review]
LGTM. We definitely don't need to care about modules that definitely trap, but it's more realistic that there might be modules that use a global as the offset, in which case it would be unfortunate if we did not optimize. (Do shared libraries already do this?)

**Comment by kripken** [comment]
Dynamic linking uses imported offsets, correct. Maybe it's better to try to avoid any regression there, good point.

We already have a flag for "zero memory initialized". If that also let us assume the segments did not trap, that would be enough here, but it is perhaps slightly awkward. (Traps never happen is the more directly relevant flag, but in general C++ can't use it.)

**Comment by tlively** [comment]
> But I think it would be more complicated to do this idea - we'd need to track split-off parts of the original segment etc

I think this should actually be a fairly localized change to `calculateRanges`.

**Comment by kripken** [comment]
Hmm, I don't see how to do that only in `calculateRanges`? At least that information needs to reach `createSplitSegments` here:

https://github.com/WebAssembly/binaryen/blob/b5938499432a1251d94e8e7b534f14919fdaf81c/src/passes/MemoryPacking.cpp#L564-L566

That is where we skip zero ranges, but in your suggestion IIUC we'd keep the last zero in the range (to ensure we still trap). So `createSplitSegments` would need to reason about which zero ranges to keep. We could mark them `.neededForTrapping = true` perhaps or such, which is not terribly hard to do, but I'm not sure it's worth it.

I checked meanwhile and this doesn't regress any use case I can find:
* Normal modules have constant offsets and they are all in bounds.
* Dynamic linking main and side modules have a non-constant offset in the big singleton segment, and we don't optimize that. (We could, actually, with some work, in which case then we'd maybe want to revisit this?)

**Comment by tlively** [comment]
It should be sufficient to have `calculateRanges` merge the last two ranges into a single nonzero range iff the segment has a possibly-trapping offset and the last range is zeroes. That will cause `createSplitSegments` to keep that data in a segment.

**Comment by kripken** [comment]
1. What if the entire segment is zeroes?
2. This may not help much. I thought the idea was to just keep the last byte (zero or not) of a segment that might trap, but in this case it might keep a lot more?

**Comment by tlively** [comment]
Basically `calculateRanges` just comes up with an alternating sequence of "isZero" ranges that are elided and turned into `memory.fill` calls as necessary and "!isZero" ranges that are kept. There's no problem if a "!isZero" range actually has zeroes in it since the ranges are the only source of truth for the subsequent splitting. So you could also emit a "!isZero" range for a single zero byte at the end instead of doing the merging like I described above. (It might even work to have a zero-length "!isZero" range; I'm not sure whether the subsequent code will handle that correctly or not.)

**Comment by kripken** [comment]
Ok, done, see last commits. This is a little more complicated than I like given the amount of benefit, but maybe it's ok, and it is definitely future-proof in that it will help us if we split more segment types later. Note that I had to also add code in another place to handle overflows for this to work.

**Review by tlively — APPROVED** [review]
LGTM!

**Pull request merged**